tools/src/main/java/com/nedap/archie/json/JSONSchemaCreator.java

```diff
@@ -0,0 +1,298 @@
+            if(containerProperty.getCardinality() != null && containerProperty.getCardinality().getLower() > 0) {
+                propertyDef.add("minItems", containerProperty.getCardinality().getLower());
+            }
+        }
+    }
+
+    private JsonObjectBuilder createPropertyDef(BmmType type) {
+
+
+        if(type instanceof BmmOpenType) {
+            return createType("object");
+            //nothing more to be done
+        } else if (type instanceof BmmSimpleType) {
+            if(isJSPrimitive(type)) {
+                return getJSPrimitive(type);
+            } else {
+                return createPolymorphicReference(type.getBaseClass());
+            }
+        } else if (type instanceof BmmContainerType) {
+
+            BmmContainerType containerType = (BmmContainerType) type;
+            return jsonFactory.createObjectBuilder()
+                .add("type", "array")
+                .add("items", createPropertyDef(containerType.getBaseType()));
+        } else if (type instanceof BmmGenericType) {
+            if(isJSPrimitive(type)) {
+                return getJSPrimitive(type);
+            } else {
+                return createPolymorphicReference(type.getBaseClass());
+            }
+
+        }
+        throw new IllegalArgumentException("type must be a BmmType, but was " + type.getClass().getSimpleName());
+
+    }
+
+    private JsonObjectBuilder createPolymorphicReference(BmmClass type) {
+
+        List<String> descendants = getAllNonAbstractDescendants( type);
+        if(!type.isAbstract()) {
+            descendants.add(BmmDefinitions.typeNameToClassKey(type.getTypeName()));
+        }
+
+        if(descendants.isEmpty()) {
+            //this is an object of which only an abstract class exists.
+            //it cannot be represented as standard json, one would think. this is mainly access control and authored
+            //resource in the RM
+            return createType("object");
+        } else if (descendants.size() > 1) {
+            JsonArrayBuilder array = jsonFactory.createArrayBuilder();
+            array.add(createRequiredArray("_type"));
+            for(String descendant:descendants) {
+                JsonObjectBuilder typePropertyCheck = createConstType(descendant);
+                JsonObjectBuilder typeCheck = jsonFactory.createObjectBuilder().add("properties", typePropertyCheck);
+
+                JsonObjectBuilder typeReference = createReference(descendant);
+                //IF the type matches
+                //THEN check the correct type from the definitions
+                JsonObjectBuilder ifObject = jsonFactory.createObjectBuilder()
+                        .add("if", typeCheck)
+                        .add("then", typeReference);
+                array.add(ifObject);
+
+            }
+
+
+            return jsonFactory.createObjectBuilder().add("allOf", array);
+        } else {
+            return createReference(BmmDefinitions.typeNameToClassKey(type.getTypeName()));
```

Comment:
Shouldn't this use the descendants type name?

Reply:
You're absolutely right. Probably doesn't fail because in most cases when this happens, descendants.get(0) is the same as type.getTypeName(). I'll fix it.